### PR TITLE
DAOS-13515 vos: use vos format to store SMD

### PR DESCRIPTION
**Pull request by wangshilong**
1. Now for MD-on-SSD, we store SMD on the external path with LMMDB which introduced some extra work for Catastrophic recovery work. extra tool is needed to parse and verify LMMDB. It will be nice to unify to use VOS format, idea is we still could store SMD on the external path but rather than use LMMDB, VOS format is still kept, this might be ok since performance is not sensitive for SMD operations.

2. Remove main thread TLS, main thread and all xstreams
will access standalone vos tls memory which is allocated by main thread.

3. move bio test to vos test since now bio ut need vos module.

Required-githooks: true
Test-nvme: auto_md_on_ssd

### CODE REVIEW

**Comment by wangshilong** [comment]
@jolivier23 PR tried to remove LMDB, bio_ut now rely on vos module, any idea how we could fix this? because we normally build bio, and then vos.

File: utils/rpms/daos.spec
Comment (brianjmurrell) [inline utils/rpms/daos.spec:1]: You need a `Release:` bump here as well as a `%changelog` (and `debian/changelog`) entry for this change.

**Review by brianjmurrell — DISMISSED** [review]
Assuming you are looking for review as this is not marked Draft.

File: utils/rpms/daos.spec
Comment (wangshilong) [reply utils/rpms/daos.spec:1]: Sure.

Comment (brianjmurrell) [reply utils/rpms/daos.spec:1]: Just a reminder, please don't resolve conversations opened by others.  Others use them to track the issues they are raising in the PR and thus they should only be resolved by the conversation owner once their issue has been resolved correctly.

**Review by brianjmurrell — COMMENTED** [review]
Packaging updates LGTM.

**Comment by wangshilong** [comment]
> Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-12251/15/execution/node/1284/log

For PMDK mode, the only failure is known as DAOS-13586, will trigger a test for md-on-ssd mode.

**Review by jolivier23 — DISMISSED** [review]

**Comment by wangshilong** [comment]
ping reviewers, CI failures are known.

File: src/include/daos_srv/vos_types.h [hunk 1]
```diff
@@ -360,8 +360,6 @@ enum {
 	VOS_IT_FOR_DISCARD	= (1 << 7),
 	/** Entry is not committed */
 	VOS_IT_UNCOMMITTED	= (1 << 8),
-	/** Iterator is for SYSDB */
-	VOS_IT_SYSDB	= (1 << 9),
 	/** Mask for all flags */
 	VOS_IT_MASK		= (1 << 10) - 1,
```
Comment (jolivier23) [inline src/include/daos_srv/vos_types.h:364]: Should this be 1 << 9 ?

File: src/vos/vos_ilog.c [hunk 2]
```diff
@@ -680,7 +680,7 @@ vos_ilog_last_update(struct ilog_df *ilog, uint32_t type, daos_epoch_t *epc)
 	D_ASSERT(epc != NULL);
 	idx = ilog_ts_idx_get(ilog);
 
-	found = vos_ts_peek_entry(idx, type, &se_entry);
+	found = vos_ts_peek_entry(idx, type, &se_entry, standalone);
```
Comment (jolivier23) [inline src/vos/vos_ilog.c:683]: I'd really like to get rid of vos standalone someday

**Review by jolivier23 — APPROVED** [review]

File: src/include/daos_srv/vos_types.h (hunk 1 above)
Comment (wangshilong) [reply src/include/daos_srv/vos_types.h]: Sure.

File: src/vos/vos_common.c [hunk 3]
```diff
@@ -125,7 +128,7 @@ vos_profile_start(char *path, int avg)
 void
 vos_profile_stop()
 {
-	struct vos_tls *tls = vos_tls_get();
+	struct vos_tls *tls = vos_tls_get(true);
```
Comment (NiuYawei) [inline src/vos/vos_common.c:131]: Minor: vos_profile_start/stop() is dead code, they could be removed.

File: src/vos/vos_common.c [hunk 4]
```diff
@@ -143,6 +146,9 @@ vos_xsctxt_get(void)
 #ifdef VOS_STANDALONE
 	return self_mode.self_xs_ctxt;
 #else
+	if (dss_tls_get() == NULL)
```
Comment (NiuYawei) [inline src/vos/vos_common.c:149]: Minor: better to add some nice comment here.

File: src/vos/vos_iterator.c [hunk 5]
```diff
@@ -141,10 +141,29 @@ nested_prepare(vos_iter_type_t type, struct vos_iter_dict *dict,
 	*cih = vos_iter2hdl(citer);
 
 out:
-	vos_dth_set(old);
+	vos_dth_set(old, !!iter->it_for_sysdb);
 	return rc;
 }
 
+bool
```
Comment (NiuYawei) [inline src/vos/vos_iterator.c:148]: s/bool/static bool

File: src/engine/srv.c [hunk 6]
```diff
@@ -1387,10 +1374,20 @@ dss_srv_init(void)
 	return rc;
 }
 
+bool
+dss_srv_shutting_down(void)
+{
+	/* from main thread */
+	if (dss_tls_get() == NULL)
+		return dss_shutting_down;
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/engine/srv.c]: Removed.

File: src/include/daos_srv/vos_types.h [hunk 7]
```diff
@@ -358,8 +360,10 @@ enum {
 	VOS_IT_FOR_DISCARD	= (1 << 7),
 	/** Entry is not committed */
 	VOS_IT_UNCOMMITTED	= (1 << 8),
+	/** Iterator is for SYSDB */
+	VOS_IT_FOR_SYSDB	= (1 << 9),
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/include/daos_srv/vos_types.h]: Removed.

File: src/vos/vos_common.c (hunk 3 above)
Comment (wangshilong) [reply src/vos/vos_common.c:131]: Sure.

File: src/vos/vos_common.c (hunk 4 above)
Comment (wangshilong) [reply src/vos/vos_common.c:149]: ok

File: src/vos/vos_iterator.c (hunk 5 above)
Comment (wangshilong) [reply src/vos/vos_iterator.c:148]: Sure

**Review by NiuYawei — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Comment by wangshilong** [comment]
@jolivier23 would you please review this again? test failures should be known on the MD-on-SSD branch without this PR.

**Review by jolivier23 — COMMENTED** [review]
I'm very sorry but I landed a patch that modifies the spec so will need to do a merge update.

**Review by jolivier23 — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Comment by wangshilong** [comment]
@jolivier23 PR should be ready, Latest CI passed though it missed MD-on-SSD testing tag, but last CI result(without your spec change) was MD-on-SSD testing, please let me know if i need re-push for a MD-on-SSD testing.

**Pull request merged**